src/message_bus.zig

```diff
@@ -751,81 +780,55 @@ const Connection = struct {
                     self.message_bus.replicas[self.peer.replica] = self;
                 }
             },
-            .client => assert(header_received.command == .request),
-            .replica => assert(header_received.command != .request),
+            .client => assert(message.header.command == .request),
+            .replica => assert(message.header.command != .request),
         }
-        assert(header_received.cluster == self.message_bus.server.cluster);
+        assert(message.header.cluster == self.message_bus.server.cluster);
 
-        if (header_received.size == @sizeOf(Header)) {
-            // If the message has no body, deliver it directly
-            self.deliver_message();
-        } else {
-            assert(header_received.size > @sizeOf(Header));
-            self.recv_body();
-        }
-    }
-
-    fn on_recv_body(
-        self: *Connection,
-        completion: *IO.Completion,
-        result: IO.RecvError!usize,
-    ) void {
-        assert(self.recv_submitted);
-        self.recv_submitted = false;
-
-        if (self.state == .shutting_down) {
-            self.maybe_close();
+        if (self.recv_progress < message.header.size) {
+            // We have not yet received the full body.
+            self.recv();
             return;
         }
-        assert(self.state == .connected);
 
-        const bytes_received = result catch |err| {
-            // TODO: maybe don't need to close on *every* error
-            log.err("error receiving body from {}: {}", .{ self.peer, err });
+        const body = message.buffer[@sizeOf(Header)..message.header.size];
+        if (!message.header.valid_checksum_body(body)) {
+            log.err("invalid checksum on body received from {}", .{self.peer});
             self.shutdown();
             return;
-        };
-
-        // No bytes received means that the peer closed the connection.
-        if (bytes_received == 0) {
-            // Skip the shutdown syscall and go straight to close.
-            self.state = .shutting_down;
-            self.maybe_close();
-            return;
         }
-        self.recv_progress += bytes_received;
 
-        if (self.recv_progress < self.incoming_message.?.header.size) {
-            // The body has not yet been fully received.
-            self.recv_body();
-        } else {
-            self.deliver_message();
+        const sector_ceil = Journal.sector_ceil(message.header.size);
+        if (self.recv_progress > message.header.size and message.header.size != sector_ceil) {
```

Comment:
in that case we would be able to stop doing the `mem.set(u8, buffer, 0);` for each message in `MessagePool`. We could instead set the buffers to `undefined` which would set the memory to an easy to recognize `0xaa` in safe build modes while being a noop in release fast/small build modes.

Reply:
We should also add a comment where we set the buffers to `undefined` because writing `0xaa` to a 2-4mb buffer could be a hidden slowdown. It's essentially adding a memcpy for every buffer.

We might also want to think about this one some more, because ideally TigerBeetle should always be built in safe mode, and we may not want the cost of the `0xaa` write in the data plane.

Reply:
> We should also add a comment if we set the buffers to `undefined` because writing `0xaa` to a 2-4mb buffer could be a hidden slowdown. It's essentially adding a memcpy for every buffer.

We're already doing this currently to 0 the buffers no?

> We might also want to think about this one some more, because ideally TigerBeetle should always be built in safe mode, and we may not want the cost of the `0xaa` write in the data plane.

I'd propose doing this only for debug builds not release safe builds then.

Reply:
> We're already doing this currently to 0 the buffers no?

We should only need to zero any sector padding after `header.size`. We shouldn't need to zero the whole buffer when handing it out or collecting it. Our checksums should catch any kind of buffer bleed related to this, on either the client side or server side?

> I'd propose doing this only for debug builds not release safe builds then.

Perfect.